File: C++/Factor_Combinations.cpp
class Solution {
public:
    vector<vector<int>> getFactors(int n) {
        vector<vector<int>> ret = solve(n, 2);
        for (vector<int> &cand : ret) {
            reverse(cand.begin(), cand.end());
        }
        return ret;
    }
    vector<vector<int>> solve(int n, int start) {
        vector<vector<int>> ret;
        for (int i = start; i <= sqrt(n); i ++) {
            if (n%i == 0) {
                ret.push_back(vector<int> { n/i, i });
                vector<vector<int>> tmp = solve(n/i, i);
                for (vector<int> cand : tmp) {
                    cand.push_back(i);
                    ret.push_back(cand);
                }
            }
        }
        return ret;
    }
};

File: Java/Factor_Combinations.java
public class Solution {
    public List<List<Integer>> getFactors(int n) {
        return solve(n, 2);
    }
    private List<List<Integer>> solve(int n, int start) {
        List<List<Integer>> ret = new LinkedList<List<Integer>>();
        for (int x = start; x <= Math.sqrt(n); x ++) {
            if (n%x == 0) {
                ret.add(new LinkedList<Integer>(Arrays.asList(x, n/x)));
                List<List<Integer>> tmp = solve(n/x, x);
                for (List<Integer> cand : tmp) {
                    ((LinkedList<Integer>)cand).addFirst(x);
                    ret.add(cand);
                }
            }
        }
        return ret;
    }
}
